pymodm/fields.py

```diff
@@ -1180,15 +1199,23 @@ def contribute_to_class(self, cls, name):
                                                     self._on_delete)
 
     def to_python(self, value):
+        doc = value
         if isinstance(value, dict):
             # Try to convert the value into our document type.
-            return self.related_model.from_document(value)
-        elif isinstance(value, self.related_model):
-            return value
+            try:
+                doc = self.related_model.from_document(value)
+            except Exception:
+                pass
+
+        if isinstance(doc, self.related_model):
+            return doc
         elif self.model._mongometa._auto_dereference:
             # Attempt to dereference the value as an id.
             dereference_id = _import('pymodm.dereference.dereference_id')
             return dereference_id(self.related_model, value)
+
+        # TODO: should we convert value to mongo here?
+        # return self.related_model._mongometa.pk.to_mongo(value)
```

Comment:
The purpose of that test is to show that you can assign the id of a referenced document directly and it will pass validation. However, the test value "1234" is not valid. My opinions:

* Convert using "to_python" (not to_mongo) here, which could yield a different result. Since we're asking for the python value, and not the mongo one, that's what's expected here.

* The test is using a bad test value. We should change it to a valid ObjectId hex string like `"58b477046e32ab215dca2b57"`.

* I think there's a separate problem with validation happening in ReferenceField. Really, the test that fails with your changes should already be failing. We should just use `to_mongo` as our validation function, like we do with [Decimal128Field](https://github.com/mongodb/pymodm/blob/ce603f5079fb602eeb2421e1bca767f4cb2bd8ac/pymodm/fields.py#L291). We don't actually need the separate `validate_related_model` function, since checking that the pk is defined is already done in `to_mongo`.

* I notice that the error message strings from `validate_related_model` and `to_mongo` are slightly different, so that may cause some tests to fail. Those tests should be updated to reflect the new error message.